atomic_reactor/utils/resource_manager.py

```diff
@@ -6,193 +6,309 @@
-        ssh_client.close()
-        return out
-
-
-class ResourceManager:
-    '''
-    Manages build resources
-    '''
-
-    def __init__(self, hosts: List[Host]):
-        self.hosts = hosts
+    def _open_ssh_session(self):
+        """
+        Create a new SSH connection and return connection object.
+        """
+        client = paramiko.SSHClient()
+        client.set_missing_host_key_policy(paramiko.AutoAddPolicy())
+        logger.debug("Open SSH connection in remote host %s", self.hostname)
+        client.connect(
+            self.hostname, username=self.username, key_filename=self.ssh_keyfile)
+        return client
+
+    @backoff.on_exception(
+        backoff.expo,
+        RETRY_ON_SSH_EXCEPTIONS,
+        factor=BACKOFF_FACTOR,
+        max_tries=MAX_RETRIES,
+        jitter=None,  # use deterministic backoff, do not apply random jitter
+        logger=logger,
+    )
+    def _ssh_run_remote_cmd(self, client, cmd: List):
+        """
+        Run a Shell command on a remote host
+
+        :return: stdout, stderr and exitcode of shell command
+        """
+        logger.debug("Try to run command %s on remote host %s", cmd, self.hostname)
+        _, stdout, stderr = client.exec_command(cmd, timeout=SSH_COMMAND_TIMEOUT)
+        return stdout, stderr, stdout.channel.recv_exit_status()
 
-    def find(self) -> List[Host]:
+    def _ssh_run_remote_cmd_to_get_json(self, client, cmd: List):
+        _, stdout, _ = self._ssh_run_remote_cmd(client, cmd)
         try:
-            resources = []
-            for host in self.hosts:
-                host.check_avail()
-                if len(host.free_slots) > 0:
-                    resources.append(host)
-            sorted_result = sorted(resources, key=lambda h: len(h.free_slots))
-            logger.debug('sorted result: %s', sorted_result)
-            return sorted_result
-        except Exception as err:
-            logger.debug(err)
-            logger.error("Unable to read hosts file")
-            return []
-
-    def obtain_free_resource(self, prid: str) -> Tuple[Host, int]:
-        availiable_hosts = self.find()
-        for host in availiable_hosts:
-            for sid in host.free_slots:
-                if host.lock(sid, prid):
-                    return host, sid
-        raise Exception('No availiable host found')
-
-
-if __name__ == '__main__':
-    my_hosts = [Host('username', 'hostname')]
-    mngr = ResourceManager(my_hosts)
-    # locking
-    free_host, slot = mngr.obtain_free_resource('unique-name-for-pipeline-run')
-
-    # Unlocking
-    free_host.unlock(slot, 'unique-name-for-pipeline-run')
+            json_stdout = json.loads(stdout)
+        except ValueError as ex:  # Not valid JSON
+            logger.warning("Json load failed. Not a valid JSON. %s", ex)
+            return None
+        except TypeError as ex:  # Not an object
+            logger.warning("Json load failed. Not an object. %s", ex)
```

Comment:
Should we log more info about the invalid stdout? Perhaps we can log even the whole stdout value. That might be convenience for us to debug.